File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/extension/RepositoryKoraExtension.java
package ru.tinkoff.kora.database.annotation.processor.extension;

import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.database.annotation.processor.DbUtils;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Objects;

public class RepositoryKoraExtension implements KoraExtension {
    private final Elements elements;
    private final Types types;

    public RepositoryKoraExtension(ProcessingEnvironment processingEnvironment) {
        this.elements = processingEnvironment.getElementUtils();
        this.types = processingEnvironment.getTypeUtils();
    }

    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror) {
        if (typeMirror.getKind() != TypeKind.DECLARED) {
            return null;
        }
        var element = this.types.asElement(typeMirror);
        if (element.getKind() != ElementKind.INTERFACE && (element.getKind() != ElementKind.CLASS || !element.getModifiers().contains(Modifier.ABSTRACT))) {
            return null;
        }

        final TypeElement typeElement;
        if (CommonUtils.findDirectAnnotation(element, DbUtils.REPOSITORY_ANNOTATION) != null) {
            typeElement = (TypeElement) this.types.asElement(typeMirror);
        } else {
            var candidates = roundEnvironment.getRootElements().stream()
                .filter(candidate -> candidate instanceof TypeElement)
                .map(candidate -> {
                    if (types.isAssignable(candidate.asType(), typeMirror)
                        && !types.isSameType(candidate.asType(), typeMirror)) {
                        if (CommonUtils.findDirectAnnotation(candidate, DbUtils.REPOSITORY_ANNOTATION) != null) {
                            return ((TypeElement) candidate);
                        } else {
                            return types.directSupertypes(candidate.asType()).stream()
                                .filter(parentType -> parentType instanceof DeclaredType)
                                .map(parentType -> ((DeclaredType) parentType))
                                .filter(parentType -> CommonUtils.findDirectAnnotation(parentType.asElement(), DbUtils.REPOSITORY_ANNOTATION) != null)
                                .findFirst()
                                .map(parentType -> ((TypeElement) parentType.asElement()))
                                .orElse(null);
                        }
                    } else {
                        return null;
                    }
                })
                .filter(Objects::nonNull)
                .toList();

            if (candidates.isEmpty()) {
                return null;
            }

            if (candidates.size() > 1) {
                throw new ProcessingErrorException("Found '%s' suitable candidates for: %s".formatted(candidates.size(), element), element);
            } else {
                typeElement = candidates.get(0);
            }
        }

        var packageElement = this.elements.getPackageOf(typeElement);
        var repositoryName = CommonUtils.getOuterClassesAsPrefix(typeElement) + typeElement.getSimpleName().toString() + "_Impl";
        var packageName = packageElement.getQualifiedName();
        return () -> {
            var repositoryElement = this.elements.getTypeElement(packageElement.getQualifiedName() + "." + repositoryName);
            if (repositoryElement == null) {
                // annotation processor will handle it
                return ExtensionResult.nextRound();
            }
            if (!CommonUtils.hasAopAnnotations(repositoryElement)) {
                return CommonUtils.findConstructors(repositoryElement, m -> m.contains(Modifier.PUBLIC)).stream().map(ExtensionResult::fromExecutable).findFirst().orElseThrow();
            }
            var aopProxy = CommonUtils.getOuterClassesAsPrefix(repositoryElement) + repositoryElement.getSimpleName() + "__AopProxy";
            var aopProxyElement = this.elements.getTypeElement(packageName + "." + aopProxy);
            if (aopProxyElement == null) {
                // aop annotation processor will handle it
                return ExtensionResult.nextRound();
            }
            return CommonUtils.findConstructors(aopProxyElement, m -> m.contains(Modifier.PUBLIC)).stream().map(ExtensionResult::fromExecutable).findFirst().orElseThrow();
        };
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/extension/RepositoryKoraExtension.kt
package ru.tinkoff.kora.database.symbol.processor.extension

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.isPublic
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.Modifier
import ru.tinkoff.kora.database.symbol.processor.DbUtils
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.CommonAopUtils.hasAopAnnotations
import ru.tinkoff.kora.ksp.common.getOuterClassesAsPrefix

@KspExperimental
class RepositoryKoraExtension(private val kspLogger: KSPLogger) : KoraExtension {
    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tags: Set<String>): (() -> ExtensionResult)? {
        if (tags.isNotEmpty()) return null
        if (type.declaration !is KSClassDeclaration) {
            return null
        }
        val declaration = type.declaration as KSClassDeclaration
        if (declaration.classKind != ClassKind.INTERFACE && (declaration.classKind != ClassKind.CLASS || !declaration.modifiers.contains(Modifier.ABSTRACT))) {
            return null
        }
        if (declaration.findAnnotation(DbUtils.repositoryAnnotation) == null) {
            return null
        }
        return lambda@{
            val packageName = declaration.packageName.asString()
            val repositoryName: String = declaration.getOuterClassesAsPrefix() + declaration.simpleName.asString() + "_Impl"
            val repositoryElement = resolver.getClassDeclarationByName("$packageName.$repositoryName")
            if (repositoryElement == null) {
                // annotation processor will handle it
                return@lambda ExtensionResult.RequiresCompilingResult
            }
            if (!hasAopAnnotations(resolver, repositoryElement)) {
                return@lambda repositoryElement.getConstructors().map { ExtensionResult.fromConstructor(it, repositoryElement) }.first()
            }

            val aopProxy = repositoryElement.getOuterClassesAsPrefix() + repositoryElement.simpleName.getShortName() + "__AopProxy"
            val aopProxyElement = resolver.getClassDeclarationByName("$packageName.$aopProxy")
            if (aopProxyElement == null) {
                return@lambda ExtensionResult.RequiresCompilingResult
            }
            val constructor = aopProxyElement.getConstructors().filter { it.isPublic() }.firstOrNull()
            if (constructor == null) {
                throw IllegalStateException()
            }
            ExtensionResult.fromConstructor(constructor, aopProxyElement)
        }
    }
}
